cos-kernel

```diff
@@ -0,0 +1,635 @@
+		list_cos_images > "${TMP_IMAGE_LIST}"
+	fi
+
+	# If we generated the list of build IDs within the past hour, use it.
+        if [[ ! -s "${TMP_BUILD_ID_FILES}" || -z "$(find "${TMP_BUILD_ID_FILES}" -cmin -60)" ]]; then
+		info getting the list of builds from "${COS_GCS_BUCKET}"
+		gsutil ls -r "${COS_GCS_BUCKET}" > "${TMP_BUILD_ID_FILES}"
+	fi
+
+	# Get and sort the list of build IDs in $COS_GCS_BUCKET.
+	if [[ -n "${BUILD_ID}" ]]; then
+		echo "${BUILD_ID}" > "${TMP_BUILD_ID_LIST}"
+		# The $BUILD_ID may be deprecated or obsolete, but
+		# becaue it was specified on the command line, we
+		# still want to print it.
+		ALL="yes"
+	else
+		grep '^gs://.*:$' "${TMP_BUILD_ID_FILES}" | \
+			grep -E '[0-9]+\.[0-9]+\.[0-9]+' | \
+			sed -e "s;${COS_GCS_BUCKET}/;;" -e "s;/:;;" | \
+			sort -V > "${TMP_BUILD_ID_LIST}"
+	fi
+
+	# Build and print the header.
+	header="BUILD_ID       MS FAMILY"
+	if [[ -n "${ALL}" ]]; then
+		header="${header} STAT"
+	fi
+	header="${header}   HDR SRC KEY TLC"
+	echo "${header}"
+
+	n=0
+	while read -r build_id; do
+		# Although we no longer create releases with the exact
+		# same build ID in different image families, there are
+		# still older releases like cos-65-10323-104-0 and
+		# cos-stable-65-10323-104-0 that do have the same
+		# build ID.  So, grep can return multiple lines.
+		all_lines=("$(grep "${build_id//./-}" "${TMP_IMAGE_LIST}")")
+		while read -r line; do
+			if [[ "${line}" == *"DEPRECATED"* && -z "${ALL}" ]]; then
+				continue
+			fi
+			milestone_family=($(get_milestone_family "${line}"))
+			printf "%-14s %2s %6s" "${build_id}" "${milestone_family[0]}" "${milestone_family[1]}"
+			if [[ -n "${ALL}" ]]; then
+				if [[ "${line}" == *"DEPRECATED"* ]]; then
+					echo -n "  dep"
+				elif [[ "${line}" == *"OBSOLETE"* ]]; then
+					echo -n "  obs"
+				else
+					echo -n "     "
+				fi
+			fi
+			echo -n "   "
+			for f in "${FILES_TO_FETCH[@]}"; do
+				if grep -q "/${build_id}/${f}\$" "${TMP_BUILD_ID_FILES}"; then
+					echo -n "+++ "
+				else
+					echo -n "--- "
+				fi
+			done
+			echo
+			n=$((n + 1))
+			if [[ "${n}" -gt 25 ]]; then
+				echo
+				echo "${header}"
+				n=0
+			fi
+		done <<< "${all_lines[@]}"
+	done < "${TMP_BUILD_ID_LIST}"
+}
+
+
+subcmd_fetch() {
+	local f
+	local md5	# md5sum checksum file
+
+	mkdir -p "${FETCHED_FILES_DIR}"
+	for f in "${FILES_TO_FETCH[@]}"; do
+		if [[ -s "${FETCHED_FILES_DIR}/${f}" ]]; then
+			"${ECHO}" "${FETCHED_FILES_DIR}/${f}" already exists
+		else
+			"${ECHO}" fetching "${COS_GCS_BUCKET}/${BUILD_ID}/${f}"
+			if ! fetch_file "${COS_GCS_BUCKET}/${BUILD_ID}/${f}" "${FETCHED_FILES_DIR}/${f}"; then
+				# Old builds do not have toolchain.
+				# Also, trusted key was introduced in R73-11391.0.0.
+				if [[ "${f}" != "${TOOLCHAIN}" && "${f}" != "${TRUSTED_KEY}" ]]; then
+					fatal could not fetch "${COS_GCS_BUCKET}/${BUILD_ID}/${f}"
+				fi
+				if [[ "${f}" == "${TRUSTED_KEY}" && "${BUILD_ID/.*/}" -gt "11391" ]]; then
```

Comment:
I guess we should tolerate trusted key download failures until we start automatically publishing the trusted key in cos-tools.

Reply:
Reworked the code.  If kernel config option CONFIG_SYSTEM_TRUSTED_KEYS is set to a key (e.g., "certs/trusted_key.pem") but we could not fetch the key, cos-kernel changes the option to  CONFIG_SYSTEM_TRUSTED_KEYS="" so that make won't fail.  It prints a message telling the user about this modification.